### PR TITLE
Reimplementation UpdateKeyViews() based in ProxyRowResponder

### PR DESCRIPTION
**Pull request by netonjm**
- [x] Fixes NRE trying BecomeFirstResponder when GetRow gets a null object (probably a race condition reloading)
- [x] Reimplements hack to set first responder based in _primitiveSetDefaultNextKeyView 

The old approach was doing a recusive iteration on every editor multiple times re-configuring next responder in any focus change, which was complicated to maintain/understand and dowhile{} was giving us unexpected results (Infinite loops).

To try improve how focus is handled over this views, now we only calculate the next/previous responder "on demand", when tab or shift+tab is pressed in a specific view implementing a ProxyRowResponder which knows if is first/end view and moves to the next available editor.

Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1472832

### CODE REVIEW

File: Xamarin.PropertyEditing.Mac/Controls/Custom/FocusableBooleanButton.cs [hunk 1]
```diff
@@ -3,7 +3,28 @@
...
+			switch (theEvent.KeyCode) {
+			case (int)NSKey.Tab:
+				if (ProxyResponder != null) {
+					if (theEvent.ModifierFlags.HasFlag(NSEventModifierMask.ShiftKeyMask)) {
+						ProxyResponder.PreviousResponder ();
+					} else {
+						ProxyResponder.NextResponder ();
+					}
+				}
+				return;
```
Comment (Therzok) [inline Xamarin.PropertyEditing.Mac/Controls/Custom/FocusableBooleanButton.cs:21]: Shouldn't this happen only when there is a proxy responder?

Comment (Therzok) [inline Xamarin.PropertyEditing.Mac/Controls/Custom/FocusableBooleanButton.cs:21]: Shouldn't this happen only when there is a proxy responder?

```suggestion
					}
					return;
				}
```

Comment (netonjm) [reply Xamarin.PropertyEditing.Mac/Controls/Custom/FocusableBooleanButton.cs:21]: true like water

File: Xamarin.PropertyEditing.Mac/Controls/Custom/FocusableComboBox.cs [hunk 2]
```diff
@@ -5,6 +5,24 @@ namespace Xamarin.PropertyEditing.Mac
...
+		{
+			switch (theEvent.KeyCode) {
+			case (int)NSKey.Tab:
+				if (ProxyResponder != null) {
+					if (theEvent.ModifierFlags.HasFlag(NSEventModifierMask.ShiftKeyMask)) {
+						ProxyResponder.PreviousResponder ();
+					} else {
+						ProxyResponder.NextResponder ();
+					}
+				}
```
Comment (netonjm) [inline Xamarin.PropertyEditing.Mac/Controls/Custom/FocusableComboBox.cs:20]: same here

File: Xamarin.PropertyEditing.Mac/Controls/Custom/FocusablePopupButton.cs [hunk 3]
```diff
@@ -8,11 +8,29 @@ internal class FocusablePopUpButton : NSPopUpButton
...
+			switch (theEvent.KeyCode) {
+			case (int)NSKey.Tab:
+				if (ProxyResponder != null) {
+					if (theEvent.ModifierFlags.HasFlag(NSEventModifierMask.ShiftKeyMask)) {
+						ProxyResponder.PreviousResponder ();
+					} else {
+						ProxyResponder.NextResponder ();
+					}
+				}
+				return;
```
Comment (netonjm) [inline Xamarin.PropertyEditing.Mac/Controls/Custom/FocusablePopupButton.cs:29]: and here

File: Xamarin.PropertyEditing.Mac/Controls/Custom/FocusableBooleanButton.cs (hunk 1 above)
Comment (netonjm) [reply Xamarin.PropertyEditing.Mac/Controls/Custom/FocusableBooleanButton.cs]: fixed

File: Xamarin.PropertyEditing.Mac/Controls/BasePathEditorControl.cs [hunk 4]
```diff
@@ -23,11 +23,35 @@ internal abstract class BasePathEditorControl<T> : PropertyEditorControl<Propert
 		private readonly NSObject[] objects;
 		public override NSObject[] AccessibilityChildren { get => this.objects; }
 
+		private readonly TextNextResponderDelegate textNextResponderDelegate;
+
+		class BasePathEditordDelegate : TextNextResponderDelegate
```
Comment (sandyarmstrong) [inline Xamarin.PropertyEditing.Mac/Controls/BasePathEditorControl.cs:28]:
```suggestion
		class BasePathEditorDelegate : TextNextResponderDelegate
```

Comment (netonjm) [reply Xamarin.PropertyEditing.Mac/Controls/BasePathEditorControl.cs:28]: Good**d** catch**d** :-)

File: Xamarin.PropertyEditing.Mac/Controls/Custom/FocusableBooleanButton.cs [hunk 5]
```diff
@@ -3,7 +3,29 @@
 
 namespace Xamarin.PropertyEditing.Mac
 {
-	internal class FocusableBooleanButton : NSButton
+	internal class ProxyResponderButton : NSButton
```
[the remark this answers is not in the document]
Comment (netonjm) [reply Xamarin.PropertyEditing.Mac/Controls/Custom/FocusableBooleanButton.cs]: fixed

**Review by BretJohnson — APPROVED** [review]

File: Xamarin.PropertyEditing.Mac/Controls/Custom/FocusableComboBox.cs [hunk 6]
```diff
@@ -5,6 +5,25 @@ namespace Xamarin.PropertyEditing.Mac
 {
 	internal class FocusableComboBox : NSComboBox
 	{
+		public RowProxyResponder ProxyResponder { get; set; }
+
+		public override void KeyDown (NSEvent theEvent)
+		{
+			switch (theEvent.KeyCode) {
+			case (int)NSKey.Tab:
```
Comment (mkrueger) [inline Xamarin.PropertyEditing.Mac/Controls/Custom/FocusableComboBox.cs:13]: for the tab keys it's better to handle the responders "insertTab:" and "insertBacktab:" instead of interfering with the keyDown event.

File: Xamarin.PropertyEditing.Mac/Controls/Custom/FocusablePopupButton.cs [hunk 7]
```diff
@@ -8,11 +8,30 @@ internal class FocusablePopUpButton : NSPopUpButton
...
+
 		public FocusablePopUpButton ()
 		{
 			Cell.LineBreakMode = NSLineBreakMode.TruncatingMiddle;
 		}
 
+		public override void KeyDown (NSEvent theEvent)
+		{
+			switch (theEvent.KeyCode) {
+			case (int)NSKey.Tab:
```
Comment (mkrueger) [inline Xamarin.PropertyEditing.Mac/Controls/Custom/FocusablePopupButton.cs:21]: responders.

File: Xamarin.PropertyEditing.Mac/Controls/Custom/ProxyResponderButton.cs [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+namespace Xamarin.PropertyEditing.Mac
+{
+	internal class ProxyResponderButton : NSButton
+	{
+		public RowProxyResponder ProxyResponder { get; set; }
+
+		public override void KeyDown (NSEvent theEvent)
+		{
+			switch (theEvent.KeyCode) {
+			case (int)NSKey.Tab:
```
Comment (mkrueger) [inline Xamarin.PropertyEditing.Mac/Controls/Custom/ProxyResponderButton.cs:12]: responders.

File: Xamarin.PropertyEditing.Mac/Controls/EntryPropertyEditor.cs [hunk 9]
```diff
@@ -2,9 +2,34 @@
...
 namespace Xamarin.PropertyEditing.Mac
 {
+	class TextNextResponderDelegate : NSTextFieldDelegate
+	{
+		public RowProxyResponder ProxyResponder { get; set; }
+
+		public override bool DoCommandBySelector (NSControl control, NSTextView textView, Selector commandSelector)
+		{
+			switch (commandSelector.Name) {
+			case "insertTab:":
```
Comment (mkrueger) [inline Xamarin.PropertyEditing.Mac/Controls/EntryPropertyEditor.cs:16]: Here you do it :).

**Review by mkrueger — CHANGES_REQUESTED** [review]

File: Xamarin.PropertyEditing.Mac/Controls/PropertyEditorControl.cs [hunk 10]
```diff
@@ -146,6 +112,66 @@ public sealed override void ViewDidChangeEffectiveAppearance ()
...
+
+					if (ctrl != null) {
+						Window?.MakeFirstResponder (ctrl.FirstKeyView);
+						return;
+					}
+				}
+			}
+		}
+
+		public void OnPreviousResponderRequested ()
```
Comment (sandyarmstrong) [inline Xamarin.PropertyEditing.Mac/Controls/PropertyEditorControl.cs:144]: A lot of duplicated code between these, maybe they could be unified a bit? Maybe both call into `OnNextResponderRequested(bool reverse)` or something?

**Review by sandyarmstrong — APPROVED** [review]
Looks good, just address Mike's stuff and maybe clean up duplication in `OnNextResponderRequested`/`OnPreviousResponderRequested`.

File: Xamarin.PropertyEditing.Mac/Controls/EntryPropertyEditor.cs [hunk 11]
```diff
@@ -2,9 +2,34 @@
...
+				if (ProxyResponder?.NextResponder () ?? false)
```
Comment (Therzok) [inline Xamarin.PropertyEditing.Mac/Controls/EntryPropertyEditor.cs:17]: You can just guarantee the ProxyResponder is not null and that will make the code readable.

Comment (Therzok) [reply Xamarin.PropertyEditing.Mac/Controls/EntryPropertyEditor.cs:17]: In example, pass it in via constructor and do `responder ?? throw new ArgumentNullException()`, that way you don't have to null check everywhere.

File: Xamarin.PropertyEditing.Mac/Controls/EntryPropertyEditor.cs [hunk 12]
```diff
@@ -2,9 +2,34 @@
 using AppKit;
 using Foundation;
 using Xamarin.PropertyEditing.ViewModels;
+using ObjCRuntime;
 
 namespace Xamarin.PropertyEditing.Mac
 {
+	class TextNextResponderDelegate : NSTextFieldDelegate
```
Comment (Therzok) [inline Xamarin.PropertyEditing.Mac/Controls/EntryPropertyEditor.cs:9]: Can we change the prefix? `TextNextResponderDelegate` doesn't make sure sense to me. `DelegatedRowTextFieldDelegate` or something based on the ProxyResponder would be nice.

File: Xamarin.PropertyEditing.Mac/Controls/Custom/ProxyResponderButton.cs [hunk 13]
```diff
@@ -0,0 +1,26 @@
+﻿using AppKit;
+
+namespace Xamarin.PropertyEditing.Mac
+{
+	internal class ProxyResponderButton : NSButton
+	{
+		public RowProxyResponder ProxyResponder { get; set; }
```
Comment (Therzok) [inline Xamarin.PropertyEditing.Mac/Controls/Custom/ProxyResponderButton.cs:7]: Inconsistency.

`ProxyResponderButton` vs `RowProxyResponder`.

File: Xamarin.PropertyEditing.Mac/Controls/Custom/NumericTextField.cs [hunk 14]
```diff
@@ -156,14 +172,16 @@ public override void DidEndEditing (NSNotification notification)
...
-	internal class KeyUpDownDelegate : NSTextFieldDelegate
+	internal class KeyUpDownDelegate : TextNextResponderDelegate
 	{
 		public event EventHandler<bool> KeyArrowUp;
 		public event EventHandler<bool> KeyArrowDown;
 
 		public override bool DoCommandBySelector (NSControl control, NSTextView textView, Selector commandSelector)
 		{
-			switch (commandSelector.Name) {
+			if (!base.DoCommandBySelector(control, textView, commandSelector))
```
Comment (Therzok) [inline Xamarin.PropertyEditing.Mac/Controls/Custom/NumericTextField.cs:182]: For a second there, I thought this called `NSTextFieldDelegate.DoCommandBySelector`, which would've thrown an exception.

File: Xamarin.PropertyEditing.Mac/Controls/Custom/NumericTextField.cs [hunk 15]
```diff
@@ -51,6 +51,22 @@ public NumericTextField ()
 			Delegate = keyUpDownDelegate;
 		}
 
+		public RowProxyResponder ProxyResponder
+		{
+			get {
+				if (Delegate is KeyUpDownDelegate keydown) {
+					return keydown.ProxyResponder;
+				}
+				return null;
```
Comment (Therzok) [inline Xamarin.PropertyEditing.Mac/Controls/Custom/NumericTextField.cs:60]:
```suggestion
				return Delegate is KeyUpDownDelegate keydown ? keydown : null;
```

File: Xamarin.PropertyEditing.Mac/Controls/CharEditorControl.cs [hunk 16]
```diff
@@ -13,7 +13,11 @@ public CharEditorControl (IHostResourceProvider hostResources)
 
 		protected override EntryPropertyEditorDelegate<char> CreateDelegate (PropertyViewModel<char> viewModel)
 		{
-			return new CharDelegate (viewModel);
+			var charDelegate = new CharDelegate (viewModel)
```
Comment (Therzok) [inline Xamarin.PropertyEditing.Mac/Controls/CharEditorControl.cs:16]: Enforce ProxyResponder as a constructor parameter.

It will help catch places where the proxy responder isn't set by having it a required parameter.

File: Xamarin.PropertyEditing.Mac/Controls/BasePathEditorControl.cs [hunk 17]
```diff
@@ -23,11 +23,35 @@ internal abstract class BasePathEditorControl<T> : PropertyEditorControl<Propert
...
+			}
+		}
+
 		protected BasePathEditorControl (IHostResourceProvider hostResource)
 			: base (hostResource)
 		{
 			this.currentTextField = new TextFieldSmallButtonContainer ();
-			this.currentTextField.Changed += CurrentTextField_Changed;
+
+			this.textNextResponderDelegate = new BasePathEditorDelegate (this)
```
Comment (Therzok) [inline Xamarin.PropertyEditing.Mac/Controls/BasePathEditorControl.cs:50]: Pass the ProxyResponder as a parameter. Or just construct it in the delegate, you have `this` anyway.

File: Xamarin.PropertyEditing.Mac/Controls/TimeSpanEditorControl.cs [hunk 18]
```diff
@@ -13,7 +13,9 @@ public TimeSpanEditorControl (IHostResourceProvider hostResources)
 
 		protected override EntryPropertyEditorDelegate<TimeSpan> CreateDelegate (PropertyViewModel<TimeSpan> viewModel)
 		{
-			return new TimeSpanDelegate (viewModel);
+			return new TimeSpanDelegate (viewModel) {
```
Comment (Therzok) [inline Xamarin.PropertyEditing.Mac/Controls/TimeSpanEditorControl.cs:16]: This could also take a constructor parameter

File: Xamarin.PropertyEditing.Mac/Controls/Custom/ProxyResponderButton.cs [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+
+		public override void KeyDown (NSEvent theEvent)
+		{
+			switch (theEvent.KeyCode) {
+			case (int)NSKey.Tab:
+				if (ProxyResponder != null) {
```
Comment (Therzok) [inline Xamarin.PropertyEditing.Mac/Controls/Custom/ProxyResponderButton.cs:13]: In example, this whole implementation makes little sense if ProxyResponder is null. Pass it in as a constructor parameter and check it's not null at that point.

File: Xamarin.PropertyEditing.Mac/Controls/PropertyEditorControl.cs [hunk 20]
```diff
@@ -146,6 +112,66 @@ public sealed override void ViewDidChangeEffectiveAppearance ()
...
+
+		public void OnNextResponderRequested ()
+		{
+			if (TableView != null) {
+				nint row = TableView.RowForView (this) + 1;
+
+				NSView view;
+				PropertyEditorControl ctrl = null;
+
+				for (; row < TableView.RowCount; row++) {
```
Comment (Therzok) [inline Xamarin.PropertyEditing.Mac/Controls/PropertyEditorControl.cs:124]: This will query `TableView.RowCount` on every loop iteration, which is an objc pinvoke. I think you want to put it above the loop:
```suggestion
				var rowCount = TableView.RowCount;
				for (; row < rowCount; row++) {
```

File: Xamarin.PropertyEditing.Mac/Controls/PropertyEditorControl.cs (hunk 10 above)
Comment (Therzok) [reply Xamarin.PropertyEditing.Mac/Controls/PropertyEditorControl.cs:144]: The random type casts and inconsistent nullability propagation usages `?.` makes it hard to read/understand.

Comment (Therzok) [reply Xamarin.PropertyEditing.Mac/Controls/PropertyEditorControl.cs:144]: Also, I agree with Sandy here, there's a lot of duplicated code that's slightly different and it's hard to follow.

File: Xamarin.PropertyEditing.Mac/Controls/ProxyRowResponder.cs [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+		{
+			this.rowType = rowType;
+			this.editorControl = new WeakReference<PropertyEditorControl> (editorControl);
+		}
+
+		public bool NextResponder()
+		{
+			if (this.editorControl.TryGetTarget (out var editor))
+			{
+				if (this.rowType == ProxyRowType.LastView || this.rowType == ProxyRowType.SingleView) {
```
Comment (Therzok) [inline Xamarin.PropertyEditing.Mac/Controls/ProxyRowResponder.cs:28]: this condition can be evaluated in the constructor :D

File: Xamarin.PropertyEditing.Mac/Controls/Custom/FocusablePopupButton.cs (hunk 7 above)
Comment (netonjm) [reply Xamarin.PropertyEditing.Mac/Controls/Custom/FocusablePopupButton.cs:21]: fixed

File: Xamarin.PropertyEditing.Mac/Controls/Custom/ProxyResponderButton.cs (hunk 8 above)
Comment (netonjm) [reply Xamarin.PropertyEditing.Mac/Controls/Custom/ProxyResponderButton.cs]: fixed

File: Xamarin.PropertyEditing.Mac/Controls/Custom/ProxyResponderButton.cs (hunk 13 above)
Comment (netonjm) [reply Xamarin.PropertyEditing.Mac/Controls/Custom/ProxyResponderButton.cs:8]: RowProxyResponder -> ProxyResponder

File: Xamarin.PropertyEditing.Mac/Controls/Custom/NumericTextField.cs (hunk 14 above)
Comment (netonjm) [reply Xamarin.PropertyEditing.Mac/Controls/Custom/NumericTextField.cs:182]: I added a comment explaining better how the base handles the commandselector

Comment (netonjm) [reply Xamarin.PropertyEditing.Mac/Controls/Custom/NumericTextField.cs:182]: fixed

File: Xamarin.PropertyEditing.Mac/Controls/PropertyEditorControl.cs (hunk 20 above)
Comment (netonjm) [reply Xamarin.PropertyEditing.Mac/Controls/PropertyEditorControl.cs]: fixed

File: Xamarin.PropertyEditing.Mac/Controls/EntryPropertyEditor.cs (hunk 12 above)
Comment (netonjm) [reply Xamarin.PropertyEditing.Mac/Controls/EntryPropertyEditor.cs]: fixed

File: Xamarin.PropertyEditing.Mac/Controls/Custom/FocusableComboBox.cs (hunk 6 above)
Comment (netonjm) [reply Xamarin.PropertyEditing.Mac/Controls/Custom/FocusableComboBox.cs:13]: insertTab: only applies to NSTextView based fields, no? 
https://developer.apple.com/documentation/appkit/nsstandardkeybindingresponding?language=objc

**Review by mkrueger — APPROVED** [review]

File: Xamarin.PropertyEditing.Mac/Controls/BasePathEditorControl.cs (hunk 17 above)
Comment (netonjm) [reply Xamarin.PropertyEditing.Mac/Controls/BasePathEditorControl.cs:50]: I want to do another review and improving it, I will include the changes in your comments on it.

File: Xamarin.PropertyEditing.Mac/Controls/CharEditorControl.cs (hunk 16 above)
Comment (netonjm) [reply Xamarin.PropertyEditing.Mac/Controls/CharEditorControl.cs:16]: I want to do another review and improve the API, for the next update I will include the changes from your comments on it.

File: Xamarin.PropertyEditing.Mac/Controls/Custom/ProxyResponderButton.cs (hunk 19 above)
Comment (netonjm) [reply Xamarin.PropertyEditing.Mac/Controls/Custom/ProxyResponderButton.cs:13]: Let's fix this small bits in the next PR changes improving the API.

File: Xamarin.PropertyEditing.Mac/Controls/Custom/NumericTextField.cs (hunk 15 above)
Comment (netonjm) [reply Xamarin.PropertyEditing.Mac/Controls/Custom/NumericTextField.cs]: fixed

**Comment by BretJohnson** [comment]
@gitbot backport main
